In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [5]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [6]:
data_path =  '/content/drive/MyDrive/ASSIGNMENT-3/Obesity Dataset.csv'
df = pd.read_csv(data_path)
df

Gender        Age    Height      Weight family_history_with_overweight  \
0     Female  21.000000  1.620000   64.000000                            yes   
1     Female  21.000000  1.520000   56.000000                            yes   
2       Male  23.000000  1.800000   77.000000                            yes   
3       Male  27.000000  1.800000   87.000000                             no   
4       Male  22.000000  1.780000   89.800000                             no   
...      ...        ...       ...         ...                            ...   
2106  Female  20.976842  1.710730  131.408528                            yes   
2107  Female  21.982942  1.748584  133.742943                            yes   
2108  Female  22.524036  1.752206  133.689352                            yes   
2109  Female  24.361936  1.739450  133.346641                            yes   
2110  Female  23.664709  1.738836  133.472641                            yes   

     FAVC  FCVC  NCP       CAEC SMOKE      CH2O  SCC       FAF       TUE  \
0      no   2.0  3.0  Sometimes    no  2.000000   no  0.000000  1.000000   
1      no   3.0  3.0  Sometimes   yes  3.000000  yes  3.000000  0.000000   
2      no   2.0  3.0  Sometimes    no  2.000000   no  2.000000  1.000000   
3      no   3.0  3.0  Sometimes    no  2.000000   no  2.000000  0.000000   
4      no   2.0  1.0  Sometimes    no  2.000000   no  0.000000  0.000000   
...   ...   ...  ...        ...   ...       ...  ...       ...       ...   
2106  yes   3.0  3.0  Sometimes    no  1.728139   no  1.676269  0.906247   
2107  yes   3.0  3.0  Sometimes    no  2.005130   no  1.341390  0.599270   
2108  yes   3.0  3.0  Sometimes    no  2.054193   no  1.414209  0.646288   
2109  yes   3.0  3.0  Sometimes    no  2.852339   no  1.139107  0.586035   
2110  yes   3.0  3.0  Sometimes    no  2.863513   no  1.026452  0.714137   

            CALC                 MTRANS           NObeyesdad  
0             no  Public_Transportation        Normal_Weight  
1      Sometimes  Public_Transportation        Normal_Weight  
2     Frequently  Public_Transportation        Normal_Weight  
3     Frequently                Walking   Overweight_Level_I  
4      Sometimes  Public_Transportation  Overweight_Level_II  
...          ...                    ...                  ...  
2106   Sometimes  Public_Transportation     Obesity_Type_III  
2107   Sometimes  Public_Transportation     Obesity_Type_III  
2108   Sometimes  Public_Transportation     Obesity_Type_III  
2109   Sometimes  Public_Transportation     Obesity_Type_III  
2110   Sometimes  Public_Transportation     Obesity_Type_III  

[2111 rows x 17 columns]

In [7]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
df['Gender']= le.fit_transform(df['Gender'])
df['family_history_with_overweight']= le.fit_transform(df['family_history_with_overweight'])
df['FAVC']= le.fit_transform(df['FAVC'])
df['CAEC']= le.fit_transform(df['CAEC'])
df['SMOKE']= le.fit_transform(df['SMOKE'])
df['SCC']= le.fit_transform(df['SCC'])
df['CALC']= le.fit_transform(df['CALC'])
df['MTRANS']= le.fit_transform(df['MTRANS'])
df['NObeyesdad']= le.fit_transform(df['NObeyesdad'])
df


Gender        Age    Height      Weight  family_history_with_overweight  \
0          0  21.000000  1.620000   64.000000                               1   
1          0  21.000000  1.520000   56.000000                               1   
2          1  23.000000  1.800000   77.000000                               1   
3          1  27.000000  1.800000   87.000000                               0   
4          1  22.000000  1.780000   89.800000                               0   
...      ...        ...       ...         ...                             ...   
2106       0  20.976842  1.710730  131.408528                               1   
2107       0  21.982942  1.748584  133.742943                               1   
2108       0  22.524036  1.752206  133.689352                               1   
2109       0  24.361936  1.739450  133.346641                               1   
2110       0  23.664709  1.738836  133.472641                               1   

      FAVC  FCVC  NCP  CAEC  SMOKE      CH2O  SCC       FAF       TUE  CALC  \
0        0   2.0  3.0     2      0  2.000000    0  0.000000  1.000000     3   
1        0   3.0  3.0     2      1  3.000000    1  3.000000  0.000000     2   
2        0   2.0  3.0     2      0  2.000000    0  2.000000  1.000000     1   
3        0   3.0  3.0     2      0  2.000000    0  2.000000  0.000000     1   
4        0   2.0  1.0     2      0  2.000000    0  0.000000  0.000000     2   
...    ...   ...  ...   ...    ...       ...  ...       ...       ...   ...   
2106     1   3.0  3.0     2      0  1.728139    0  1.676269  0.906247     2   
2107     1   3.0  3.0     2      0  2.005130    0  1.341390  0.599270     2   
2108     1   3.0  3.0     2      0  2.054193    0  1.414209  0.646288     2   
2109     1   3.0  3.0     2      0  2.852339    0  1.139107  0.586035     2   
2110     1   3.0  3.0     2      0  2.863513    0  1.026452  0.714137     2   

      MTRANS  NObeyesdad  
0          3           1  
1          3           1  
2          3           1  
3          4           5  
4          3           6  
...      ...         ...  
2106       3           4  
2107       3           4  
2108       3           4  
2109       3           4  
2110       3           4  

[2111 rows x 17 columns]

In [8]:
X=df.drop(['NObeyesdad'], axis=1)
X
Y=df['NObeyesdad']
Y

0       1
1       1
2       1
3       5
4       6
       ..
2106    4
2107    4
2108    4
2109    4
2110    4
Name: NObeyesdad, Length: 2111, dtype: int64

In [9]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=41)

In [10]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree

model1=LogisticRegression(random_state=1)
model2=GaussianNB()
model3=tree.DecisionTreeClassifier(random_state=1)

In [11]:
from sklearn.ensemble import VotingClassifier
e_model= VotingClassifier(estimators=[('LR',model1),('NB',model2),('DT',model3)], voting='soft')

In [12]:
[c.fit(x_train,y_train) for c in (model1,model2,model3,e_model)]

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

[LogisticRegression(random_state=1),
 GaussianNB(),
 DecisionTreeClassifier(random_state=1),
 VotingClassifier(estimators=[('LR', LogisticRegression(random_state=1)),
                              ('NB', GaussianNB()),
                              ('DT', DecisionTreeClassifier(random_state=1))],
                  voting='soft')]

In [13]:
from sklearn import metrics
y_pred= model1.predict(x_test)
print('MODEL 1 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model2.predict(x_test)
print('MODEL 2 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model3.predict(x_test)
print('MODEL 3 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= e_model.predict(x_test)
print('e_model ACC', metrics.accuracy_score(y_test,y_pred))

MODEL 1 ACC 0.6643026004728132
MODEL 2 ACC 0.6453900709219859
MODEL 3 ACC 0.9456264775413712
e_model ACC 0.9078014184397163


In [14]:
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree

model1=LogisticRegression(random_state=1)
model2=KNeighborsClassifier(n_neighbors=1,weights='uniform',algorithm='auto',leaf_size=30,
                                             p=2,metric='minkowski',metric_params=None,n_jobs=None)
model3=tree.DecisionTreeClassifier(random_state=1)

from sklearn.ensemble import VotingClassifier
e_model= VotingClassifier(estimators=[('LR',model1),('NB',model2),('DT',model3)], voting='soft')

[c.fit(x_train,y_train) for c in (model1,model2,model3,e_model)]


from sklearn import metrics
y_pred= model1.predict(x_test)
print('MODEL 1 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model2.predict(x_test)
print('MODEL 2 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model3.predict(x_test)
print('MODEL 3 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= e_model.predict(x_test)
print('e_model ACC', metrics.accuracy_score(y_test,y_pred))

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

MODEL 1 ACC 0.6643026004728132
MODEL 2 ACC 0.8959810874704491
MODEL 3 ACC 0.9456264775413712
e_model ACC 0.9361702127659575


In [24]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree

model1=RandomForestClassifier()
model2=KNeighborsClassifier(n_neighbors=1,weights='uniform',algorithm='auto',leaf_size=30,
                                             p=2,metric='minkowski',metric_params=None,n_jobs=None)
model3=tree.DecisionTreeClassifier(random_state=1)

from sklearn.ensemble import VotingClassifier
e_model= VotingClassifier(estimators=[('RF',model1),('NB',model2),('DT',model3)], voting='hard')

[c.fit(x_train,y_train) for c in (model1,model2,model3,e_model)]


from sklearn import metrics
y_pred= model1.predict(x_test)
print('MODEL 1 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model2.predict(x_test)
print('MODEL 2 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= model3.predict(x_test)
print('MODEL 3 ACC', metrics.accuracy_score(y_test,y_pred))

y_pred= e_model.predict(x_test)
print('e_model ACC', metrics.accuracy_score(y_test,y_pred))

MODEL 1 ACC 0.9598108747044918
MODEL 2 ACC 0.8959810874704491
MODEL 3 ACC 0.9456264775413712
e_model ACC 0.966903073286052


                                                          SCIKIT IMPLEMENTATION

In [16]:
X=df.drop(['NObeyesdad'], axis=1)
X
Y=df['NObeyesdad']
X

Gender        Age    Height      Weight  family_history_with_overweight  \
0          0  21.000000  1.620000   64.000000                               1   
1          0  21.000000  1.520000   56.000000                               1   
2          1  23.000000  1.800000   77.000000                               1   
3          1  27.000000  1.800000   87.000000                               0   
4          1  22.000000  1.780000   89.800000                               0   
...      ...        ...       ...         ...                             ...   
2106       0  20.976842  1.710730  131.408528                               1   
2107       0  21.982942  1.748584  133.742943                               1   
2108       0  22.524036  1.752206  133.689352                               1   
2109       0  24.361936  1.739450  133.346641                               1   
2110       0  23.664709  1.738836  133.472641                               1   

      FAVC  FCVC  NCP  CAEC  SMOKE      CH2O  SCC       FAF       TUE  CALC  \
0        0   2.0  3.0     2      0  2.000000    0  0.000000  1.000000     3   
1        0   3.0  3.0     2      1  3.000000    1  3.000000  0.000000     2   
2        0   2.0  3.0     2      0  2.000000    0  2.000000  1.000000     1   
3        0   3.0  3.0     2      0  2.000000    0  2.000000  0.000000     1   
4        0   2.0  1.0     2      0  2.000000    0  0.000000  0.000000     2   
...    ...   ...  ...   ...    ...       ...  ...       ...       ...   ...   
2106     1   3.0  3.0     2      0  1.728139    0  1.676269  0.906247     2   
2107     1   3.0  3.0     2      0  2.005130    0  1.341390  0.599270     2   
2108     1   3.0  3.0     2      0  2.054193    0  1.414209  0.646288     2   
2109     1   3.0  3.0     2      0  2.852339    0  1.139107  0.586035     2   
2110     1   3.0  3.0     2      0  2.863513    0  1.026452  0.714137     2   

      MTRANS  
0          3  
1          3  
2          3  
3          4  
4          3  
...      ...  
2106       3  
2107       3  
2108       3  
2109       3  
2110       3  

[2111 rows x 16 columns]

In [17]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=41)

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree

In [19]:
model1=LogisticRegression(random_state=1)
model2=KNeighborsClassifier(n_neighbors=1,weights='uniform',algorithm='auto',leaf_size=30,
                                             p=2,metric='minkowski',metric_params=None,n_jobs=None)
model3=tree.DecisionTreeClassifier(random_state=1)

In [20]:
estimators=[('LR',model1),('KNN',model2),('DT',model3)]

In [21]:
from sklearn.ensemble import RandomForestClassifier
final_model=RandomForestClassifier()
final_estimator=final_model

In [22]:
from sklearn.ensemble import StackingClassifier
stacked_clf= StackingClassifier(estimators=estimators, final_estimator=final_estimator)
stacked_clf.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

StackingClassifier(estimators=[('LR', LogisticRegression(random_state=1)),
                               ('KNN', KNeighborsClassifier(n_neighbors=1)),
                               ('DT', DecisionTreeClassifier(random_state=1))],
                   final_estimator=RandomForestClassifier())

In [23]:
y_pred=stacked_clf.predict(x_test)
print('ACC', metrics.accuracy_score(y_test, y_pred))

ACC 0.9479905437352246


**RESULT & DISCUSSION **:- Here, In a obesity dataset, the ensemble technique is applied to classify the type of obesity. At first, Logistics Regression, Naive Bayes and Decision Tree algorithms are combined. The Ensemble model accuracy is 0.90 whether first 3 model's accuracy are respectively 0.66,0.64,0.94. So,the result is not as expexted. Then, Logistics Regression, KNN and Decision Tree algorithms are combined. The Ensemble model accuracy is 0.93 whether first 3 model's accuracy are respectively 0.66,0.89,0.94. So,this result is also not as expexted. After that, Random Forest, Naive Bayes and Decision Tree algorithms are combined. The Ensemble model accuracy is 0.96 whether first 3 model's accuracy are respectively 0.95,0.89,0.94. So,At last, the optimum result comes out. We got the best accuracy. Here, in the first two trial, soft ensemble technique is used. But in the last trial, to get the optimum result ensemble technique is changed to the hard technique. Here, We also used the advanced ensemble techniques that is stacking. In this stacking technique, the accuracy we got is 0.94, though it is less than before hard/soft ensemble technique.I think we can improve this accuracy by using different classifier or by doing data-pre-processing cautiously etc.

**Strength of Ensemble ** :- Ensemble methods often outperform individual models by combining the predictions of multiple base models. This can lead to higher accuracy and better generalization on unseen data.By combining multiple models, ensemble methods can mitigate overfitting. Ensemble methods are flexible and can incorporate a wide range of base models, including classifiers, regressors, and even more complex models like neural networks. This versatility allows ensemble techniques to be applied to various types of problems across different domains.

**Weakness of Ensemble** :- Stacking involves training multiple base models and a meta-model, which can increase computational overhead, especially when dealing with large datasets or complex models. Ensemble techniques often involve tuning multiple hyperparameters for each base model and the meta-model. Finding the optimal hyperparameters can be challenging and may require extensive experimentation, increasing the overall complexity of the model selection process.



                                                                            
                                                                             
                                                                  **THE END**